src/persistent_backend_manager.cc

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) 2023, NVIDIA CORPORATION. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions
+// are met:
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of NVIDIA CORPORATION nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "persistent_backend_manager.h"
+
+#include <memory>
+#include <mutex>
+#include "backend_config.h"
+#include "filesystem/api.h"
+#include "triton/common/logging.h"
+
+namespace triton { namespace core {
+
+static std::weak_ptr<PersistentBackendManager> persist_backend_manager_;
+static std::mutex mu_;
+
+Status
+PersistentBackendManager::Create(
+    const triton::common::BackendCmdlineConfigMap& config_map,
+    std::shared_ptr<PersistentBackendManager>* manager,
+    std::shared_ptr<TritonBackendManager>& backend_manager)
+{
+  std::lock_guard<std::mutex> lock(mu_);
+
+  // If there is already a manager then we just use it...
+  *manager = persist_backend_manager_.lock();
+  if (*manager != nullptr) {
+    return Status::Success;
+  }
+
+  manager->reset(new PersistentBackendManager());
+  persist_backend_manager_ = *manager;
+
+  RETURN_IF_ERROR(
+      (*manager)->InitPersistentBackends(config_map, backend_manager));
+  return Status::Success;
+}
+
+Status
+PersistentBackendManager::InitPersistentBackends(
+    const triton::common::BackendCmdlineConfigMap& config_map,
+    std::shared_ptr<TritonBackendManager>& backend_manager)
+{
+  for (const auto& be : {"pytorch"}) {
+    RETURN_IF_ERROR(InitPersistentBackend(be, config_map, backend_manager));
+  }
+
+  return Status::Success;
+}
+
+Status
+PersistentBackendManager::InitPersistentBackend(
```

Comment:
To follow on @dzier's suggestions - if we just implemented this method alone and used it to initialize pytorch before other models - does that work as well? We could avoid the mutex, singleton and keep a static method: PreloadBackend()

Reply:
We still need the mutex since the user can create multiple TritonServer objects in the C-API and we want to make sure that this library is only dlopened once: https://github.com/triton-inference-server/server/blob/main/qa/L0_multi_server/test.sh